Add num_measurements argument to unshard() and decode_result().

MxmUrw opened the pull request:
I implemented here what @divergentdave suggested in #256.

Currently this is only the `num_measurements` parameter being passed to the `decode_result()` function of prio3 types via `Collector::unshard()`.

Is there something else to be done? For example add an average type like was done in cfrg/draft-irtf-cfrg-vdaf#95?

divergentdave reviewed (changes requested):
Thanks! If you would like to add an average type parallel to that in the Sage implementation, that should go in a unit test, in one of the `tests` modules throughout the tree. These changes need a pass through `cargo fmt`, but other than that, the CI should be good.

MxmUrw commented on src/vdaf.rs:
```diff
@@ -210,6 +210,7 @@ where
         &self,
         agg_param: &Self::AggregationParam,
         agg_shares: M,
+        num_shares: usize,
```
Yes, that's true. Thanks for catching.

MxmUrw commented:
Then I'm also going to add that average type.

MxmUrw commented:
I added the average type. Some notes:
1. To have less code duplication, I kept it as minimal as possible (for instance the validity circuit always accepts).
2. It seemed difficult to put all code into a single "self contained" file because of visibility restrictions (`run_vdaf` cannot be used in the `tests` directory though I do not quite understand why; the fields of `Prio3` are private to the module). So I mostly followed the structure of how the other prio3 types are defined and tested.

cjpatton commented on src/flp/test_types.rs:
```diff
@@ -0,0 +1,112 @@
+use crate::field::FieldElement;
```
I believe this needs the copyright notice, as seen in other files.
```
// SPDX-License-Identifier: MPL-2.0
```

cjpatton commented on src/vdaf/prio3.rs:
```diff
@@ -159,6 +162,34 @@ impl Prio3Aes128Histogram {
+/// aggregate is the arithmetic average. This type is a test for demonstrating that the number of measurements
+/// may be used in the `decode_result()` function. No validity checks are done on the input data.
+#[cfg(test)]
+pub type TestPrio3Aes128Average = Prio3<Average<Field128>, PrgAes128, 16>;
```
If this type is indeed only for testing, then it need not be `pub`; `pub(crate)` would be sufficient. However I think it might be worth exposing in the public API. In this case, we could drop the `Test` prefix.

cjpatton commented on src/flp/test_types.rs:
```diff
@@ -0,0 +1,112 @@
+    type AggregateResult = f64;
+    type Field = F;
+
+    fn encode_measurement(&self, summand: &F::Integer) -> Result<Vec<F>, FlpError> {
```
It should be possible to refactor `Average` and `Sum` to reuse code. In particular:
* encode_measurement()
* decode_result()
* truncate()
* valid()

all have lots of code in common. (There might be others, too.)

cjpatton commented on src/flp/test_types.rs:
```diff
@@ -0,0 +1,112 @@
+
+impl<F: FieldElement> Type for Average<F> {
+    type Measurement = F::Integer;
+    type AggregateResult = f64;
```
As I mentioned [here](https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/95#discussion_r916372075), I'm a bit wary of having the result type (`f64`) have a different precision than the measurements (`u64`, `u128`, ...). Maybe this concern is overblown?

MxmUrw replied:
I do not quite understand the direction that you would like to go with this. It doesn't seem to be about possible loss of information, as in your alternative information is also lost in the `floor()`. Is this about floats being more "lossy" when dealing with larger numbers?

I think that if this (or sth. similar) is the concern, then it is going to depend on the concrete situation where such an average type is going to be used, and the trade-offs cannot really be decided a priori. Alternatively, it could be that something else is going on here that I don't see?

This is also interesting for our bounded-norm fixed-point vector type because there we currently also use floats. In particular, in our use-case, we are going to have a `float -> fixed-point (prio) -> float` pipeline in some form, simply because ML frameworks use floats.

cjpatton replied:
Good point about information loss. And yeah, basically my concern is: When one is given a float type, they expect some imprecision inherent to the IEEE standard for floating point arithmetic. But if the measurements are integers (`u64` or `u128` depending on the field being used to represent them), one would expect to get full precision from the result. Put another way: `Average` is a fairly straightforward extension of `Sum`, but is strictly less useful because it loses precision. 

> This is also interesting for our bounded-norm fixed-point vector type because there we currently also use floats in our aggregate-result-type. In particular, in our use-case, we are going to have a `float -> fixed-point (prio) -> float` pipeline in some form, simply because ML frameworks use floats.

This sounds totally reasonable! One way to solve this might be to have the measurement type for `Average` be `f64` so that the measurement type matches the aggregate result, precisely as you would for your new type.

cjpatton commented on src/flp/test_types.rs:
```diff
@@ -0,0 +1,112 @@
+        _joint_rand: &[F],
+        _num_shares: usize,
+    ) -> Result<F, FlpError> {
+        // Accept everything since this is only a test type.
```
This looks to me like a critical vulnerability waiting too happen. Imagine someone comes along and enablers `Prio3Aes128Average` without thoroughly checking the code, or uses `Average` for some other purpose (it's in the public API afterall).

In my view, this `Type` should be implemented properly, especially those in the public API. It shouldn't be too bad if we refactor the code to crib off of `Sum`.

cjpatton reviewed (changes requested):
Thanks, @MxmUrw, this is great. I have one blocking comment, which is to implement input validation for the `Average` type properly. See in-line comment for details.

cjpatton commented:
> Thanks, @MxmUrw, this is great. I have one blocking comment, which is to implement input validation for the `Average` type properly. See in-line comment for details.

Another way to resolve this: I don't think that adding code to test the new code changes is strictly necessary, given that you plan to send a PR for adding a VDAF that will exercise the code changes.

MxmUrw commented:
Thanks for the feedback! If it would be useful to have an average type in the public API, then of course I can implement it properly.

> Another way to resolve this: I don't think that adding code to test the new code changes is strictly necessary, given that you plan to send a PR for adding a VDAF that will exercise the code changes.

Actually, since code reuse is going to be a similar concern when adding our fixed point type, I think I might as well tackle that problem already here. I will try it out and push my proposal.

MxmUrw commented:
Regarding the measurement type of `Average`: If we are going to allow non-integral numbers as inputs, we are going to have a lot more to think about: These inputs would need to be encoded as (un-)signed fixed point numbers, which requires us (or the user of the type) to choose a range and a number of bits for the fractional part. These choices again depend heavily on the concrete application.

But these are all problems that needn't necessarily be dealt with by prio. The actual functionality that is needed for computing the average is just the `Sum` type (and knowing the number of clients). Everything else is pre- and postprocessing and might best be done when there is a concrete use-case.

Continuing in this spirit, I think the result type of `Average` could be a tuple `(aggregate_sum: F::Integer, num_of_measurements: usize)`, and the user of that type can decide later how he wants to deal with loss of information.

cjpatton commented:
> Regarding the measurement type of `Average`: If we are going to allow non-integral numbers as inputs, we are going to have a lot more to think about: These inputs would need to be encoded as (un-)signed fixed point numbers, which requires us (or the user of the type) to choose a range and a number of bits for the fractional part. These choices again depend heavily on the concrete application.
> 
> But these are all problems that needn't necessarily be dealt with by prio. The actual functionality that is needed for computing the average is just the `Sum` type (and knowing the number of clients). Everything else is pre- and postprocessing and might best be done when there is a concrete use-case.
> 
> Continuing in this spirit, I think the result type of `Average` could be a tuple `(aggregate_sum: F::Integer, num_of_measurements: usize)`, and the user of that type can decide later how he wants to deal with loss of information.

Yeah good point. Let's not open up this can of beans with this change. Let's just stick with returning `f64`, as you currently do.

cjpatton commented on src/vdaf/prio3_test.rs:
```diff
@@ -149,6 +149,19 @@ fn test_vec_prio3_sum() {
     }
 }
 
+#[test]
```
This test should be removed. It would only be appropriate if we had test vectors `Prio3Aes128Average`, but we don't (yet!).

cjpatton reviewed (changes requested):
Great work! Just a few comments. One high-level thing:  Can we add some basic tests to `flp/types.rs` that exercise the new `Average` type? We don't need as much test coverage as `test_sum()`, but we should at least have tests that validate any code that is different. In particular I'm thinking of `.decode_result()`.

tgeoghegan commented on src/flp/types.rs:
```diff
@@ -116,26 +120,12 @@ impl<F: FieldElement> Sum<F> {
-            )));
-        }
-
+        let bits_int = valid_int_from_usize::<F>(bits)?;
```
As a stylistic note, I agree with Chris that in this project we prefer type annotations over the turbofish.

tgeoghegan commented on src/vdaf/prio3.rs:
```diff
@@ -218,7 +243,7 @@ impl Prio3Aes128Histogram {
 ///
 /// // Unshard
-/// let agg_res = vdaf.unshard(&(), agg_shares).unwrap();
+/// let agg_res = vdaf.unshard(&(), agg_shares, 5).unwrap();
```
nit: to make this more clear, how about changing the for loop above to
```
let measurements = [0, 1, 1, 1, 0];
for measurement in measurements  { ... }
```
... and then ...
```suggestion
/// let agg_res = vdaf.unshard(&(), agg_shares, measurements.len()).unwrap();
```
I think that helps make it clear what this magic value `5` is, especially for people reading this example on docs.rs and not in an editor that can tell them what the third argument to `vdaf.unshard` is.

tgeoghegan commented on src/flp/types_utilities.rs:
```diff
@@ -0,0 +1,142 @@
+/// * `input` - The field element to encode
+/// * `bits` - The number of bits to use for the encoding
+/// * `max_input` - The maximal number representable with `bits` bits. Equal to `bits^2-1`
+pub(crate) fn encode_into_bitvector_representation<F: FieldElement>(
```
This function doesn't really operate on any `Type`, so I'd say maybe it should be put into `field.rs`. I also think you could refactor it into an associated function on trait `FieldElement` by inserting the following function definition into `pub trait FieldElement`:
```
    /// Encode `input` as `bits`-bit vector of elements of `T::Field` if it's small enough
    /// to be represented with that many bits.
    ///
    /// # Arguments
    ///
    /// * `input` - The field element to encode
    /// * `bits` - The number of bits to use for the encoding
    /// * `max_input` - The maximal number representable with `bits` bits. Equal to `bits^2-1`
    fn encode_into_bitvector_representation(
        input: &Self::Integer,
        bits: usize,
        max_input: Self::Integer,
    ) -> Result<Vec<Self>, FieldError> {
        if *input > max_input {
            return Err(FieldError::InputSizeMismatch);
        }

        let one = Self::Integer::from(Self::one());
        let mut encoded = Vec::with_capacity(bits);
        for l in 0..bits {
            let l = Self::Integer::try_from(l).unwrap();
            let w = Self::from((*input >> l) & one);
            encoded.push(w);
        }
        Ok(encoded)
    }
```

I think this makes it easier for users of the library to discover these functions, since they'll now show up when they read the documentation on [trait `FieldElement`](https://docs.rs/prio/latest/prio/field/trait.FieldElement.html), alongside the other default trait functions provided there.

I suggest the same approach for `valid_int_from_usize`, `decode_from_bitvector_representation`, `call_gadget_on_vec_entries`, `decode_result` and `decode_result_vec`.

tgeoghegan commented on src/flp/types_utilities.rs:
```diff
@@ -0,0 +1,142 @@
+}
+
+/// Check whether `input` and `joint_rand` have the length expected by `typ`.
+pub(crate) fn valid_call_check<T: Type>(
```
Similar to my suggestion of adding default implementations to `trait FieldElement`, I think this function and `truncate_call_check` could be refactored into methods on `trait Type` in `flp.rs`. In this case, I argue for _methods_ and not associated functions since these two new functions already take `typ: &T` and so that could become `&self`. e.g:

```
    /// Check whether `input` and `joint_rand` have the length expected by `self`.
    fn valid_call_check(
        &self,
        input: &[Self::Field],
        joint_rand: &[Self::Field],
    ) -> Result<(), FlpError> {
        if input.len() != self.input_len() {
            return Err(FlpError::Valid(format!(
                "unexpected input length: got {}; want {}",
                input.len(),
                self.input_len(),
            )));
        }

        if joint_rand.len() != self.joint_rand_len() {
            return Err(FlpError::Valid(format!(
                "unexpected joint randomness length: got {}; want {}",
                joint_rand.len(),
                self.joint_rand_len()
            )));
        }

        Ok(())
    }
```

tgeoghegan commented on src/flp/types.rs:
```diff
@@ -183,27 +160,109 @@ impl<F: FieldElement> Type for Sum<F> {
+
+    fn decode_result(&self, data: &[F], num_measurements: usize) -> Result<f64, FlpError> {
+        // Compute the average from the aggregated sum.
+        let data = decode_result(data).unwrap();
```
I agree with Chris that we should try our best not to panic, since (1) it's quite rude for libraries to crash and (2) this function is already wired up to return an error.

MxmUrw commented on src/flp/types.rs:
```diff
@@ -183,27 +160,109 @@ impl<F: FieldElement> Type for Sum<F> {
+        // Compute the average from the aggregated sum.
+        let data = decode_result(data).unwrap();
+        let data: u64 = data.try_into().unwrap();
+        let result = (data as f64) / (num_measurements as f64);
```
Somehow I cannot get your suggestion to work.
 - I think that `data` needs to be a `f64` here, because division seems only to be implemented in the case where both numerator and denominator have the same type. And I assume that we want to use the floating point division here.
 - There is no `from` or `try_from` for `u64 -> f64`, only for `u32 -> f64`, so I think the only other option is to use the "cast with `as`".
 - Similarly there is no `from` for `usize -> f64`.

Could you elaborate on how your comment was meant? Am I missing imports?

cjpatton replied:
Ok, this looks fine as is.

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,55 @@ pub trait FieldElement:
         Ok(vec)
     }
+
+    /// Encode `input` as `bits`-bit vector of elements of `Self` if it's small enough
```
I'm a bit wary of leaking these methods in the public API, but since @tgeoghegan suggested it feel free to ignore me :)
* encode_into_bitvector_representation
* decode_from_bitvector_representation
* valid_integer_try_from

tgeoghegan replied:
Not a bad point there, @cjpatton. The goal here was to put these new methods on all implementations of `FieldElement` without having them be floating functions in other modules. If we want to hide them from clients outside the crate, we could move them to some new trait `FieldElementExt` that is `pub(crate)`. I don't want to make Mxm do that though since this PR has been hanging out long enough as it is.

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,55 @@ pub trait FieldElement:
+    ///
+    /// * `input` - The field element to encode
+    /// * `bits` - The number of bits to use for the encoding
+    /// * `max_input` - The maximal number representable with `bits` bits. Equal to `bits^2-1`
```
`max_input` is used as a function parameter because precomputing it is convenient. If chosen incorrectly (i.e., if `max_input < 2^bits - 1`, then this method will panic in a way that the user may not expect. This was fine when this method was crate-private, but I'm a bit wary of putting this in the public API.

The fundamental goal here is to ensure that `input` can be encoded with `bits` bits. I wonder if we can rewrite the function to enforce this some other way? One straightforward approach is to compute `max_input` ourselves, but we can probably do even better in terms of computational efficiency.

cjpatton replied:
Great work!

MxmUrw commented:
Hi! Thanks again for the detailed comments.

We have dealt with all of them, except that we need input with regards to the following things:
1. Our "utility functions" now live in `flp.rs` and `field.rs`, as suggested. Which means that for example `encode_into_bitvector_representation` (which is defined in `field.rs`) throws a `FieldError`, and not an `FlpError::Encode`. To this end we also added the new constructors `FieldError::IntegerTryFrom` and `FieldError::IntegerTryInto`. This might, or might not be how this situation should be dealt with.
2. Regarding the `as f64` cast, I added a comment [above](https://github.com/divviup/libprio-rs/pull/258#discussion_r931213308).

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,55 @@ pub trait FieldElement:
+
+    /// Decode the bitvector-represented value `input` into a simple representation as a single field element.
+    /// This function panics if `2 ^ input.len()` does not fit into the field `Self`.
+    fn decode_from_bitvector_representation(input: &[Self]) -> Self {
```
Now that this method is in the public API, I think it's best to err rather than panic.

MxmUrw replied:
Yep, we just did this.

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,55 @@ pub trait FieldElement:
+    fn valid_integer_try_from<N>(i: N) -> Result<Self::Integer, FieldError> where Self::Integer: TryFrom<N>{
+        let i_int = Self::Integer::try_from(i).map_err(|_| {FieldError::IntegerTryFrom})?;
+
+        if Self::modulus() >> i_int == Self::Integer::from(Self::zero()) {
```
This doesn't look right to me: `p >> i` means `p` shifted to the right by `i` bits. Thus this condition holds if `p <= 2^i`. What we want is `p >= i`, right? (Go look at where this function is used. If what we really want is `p <= 2^i`, then the documentation needs to be updated.

cjpatton commented on src/flp.rs:
```diff
@@ -490,6 +494,46 @@ pub trait Type: Sized + Eq + Clone + Debug {
+        input: &[Self::Field],
+        joint_rand: &[Self::Field],
+    ) -> Result<(), FlpError> {
+        if input.len() != self.input_len() {
```
nit: You could call `truncate_call_check()` instead in order to de-duplicate code.

MxmUrw replied:
Yes, though the problem would be that `truncate_call_check()` returns an `FlpError::Truncate` while `valid_call_check()` returns an `FlpError::Valid`.

We could do either:
- Choose another error type which we throw in both functions.
- Call `truncate_call_check`, then do `map_err` to change the error type.

cjpatton replied:
Ack, good catch. You can leave as is.

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,55 @@ pub trait FieldElement:
+    }
+
+    /// Interpret `i` as [`Self::Integer`] if it's representable in that type and smaller than the field modulus.
+    fn valid_integer_try_from<N>(i: N) -> Result<Self::Integer, FieldError> where Self::Integer: TryFrom<N>{
```
Consider removing the `valid_` prefix here. It' doesn't make sense in this context.

MxmUrw replied:
The idea was that this is not simply a value of type `Field::Integer` but one that is smaller than the modulus, thus a "representable integer" or "valid integer". But we can remove that prefix.

cjpatton replied:
Ack, you can leave as is.

cjpatton reviewed (commented):
Nearly there!

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,72 @@ pub trait FieldElement:
+        input: &Self::Integer,
+        bits: usize,
+    ) -> Result<Vec<Self>, FieldError> {
+        // Create a mutable copy of `input`.
+        // In each iteration of the following loop
+        // we take the least significant bit,
+        // and shift input to the right by one bit.
```
nit: Block comments should break at 100 characters
```suggestion
        // Create a mutable copy of `input`. In each iteration of the following loop we take the
        // least significant bit, and shift input to the right by one bit.
```

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,72 @@ pub trait FieldElement:
+    /// Decode the bitvector-represented value `input` into a simple representation as a single field element.
+    /// This function errors if `2^input.len() - 1` does not fit into the field `Self`.
+    fn decode_from_bitvector_representation(input: &[Self]) -> Result<Self, FieldError> {
+
```
nit: Remove spurios newline

cjpatton commented on src/field.rs:
```diff
@@ -190,6 +196,72 @@ pub trait FieldElement:
+    }
+
+    /// Check if the largest number representable with `bits` bits (i.e. 2^bits - 1) is representable in this field.
+    fn valid_integer_bitlength(bits: usize) -> Result<(), FieldError> {
```
I think it would be better if this method just returned a `bool` indicating whether the `2^bits-1` is encodable and let the caller return an error if not.

cjpatton reviewed (commented):
LGTM! A couple of asks before we can merge:
* You will need to rebase your change. The following steps should suffice:
    1. You'll want to start by "squashing" your changes into a single commit.
        a. git branch backup      # First make sure your branch is backed up in case something goes wrong.
        b. git rebase main -i    # This drops you into an edit or with instructions on how to proceed. Basically your goal objective is to "squash" every commit into the previous one, except for the first, which you will "pick".
        c. git diff backup     # Confirm that your squashed branch hasn't changed
        d. git push -f   # Update your origin with the new branch. (BE CAREFUL, this rewrites your history.)
    2. Now that you have one commit with all of your changes, you''ll want to rebase on remote.  It looks like your changes on your fork of `main`, so you'll need to do something a bit funky. (Tip: In the future, when making a PR you should always give your branch a distinctive name.)
        a. git remote add upstream https://github.com/divviup/libprio-rs      # you may need to change this, depending on your setup.
        b. git fetch upstream      # Get changes from upstream divviup/libprio-rs
        c. git rebase upstream/main -i      # Rebase on main.
* Tests are currently failing. (See the CI builds at the end of the PR for information.) You should be able to reproduce locally via `cargo test --all-features` once you've properly rebased.

MxmUrw commented:
Great! Thanks for the detailed instructions. I did two force-pushes now, once after squashing and once after rebase on main + fixing the failing test, I hope that is correct.

cjpatton commented:
Rebase looks good! I'm taking a last pass to make sure. (Sorry it took me so long to get back to this.) Meanwhile, you've got one more thing to fix before we can merge. You guessed it, it's clippy time:
```
$ cargo clippy
    Checking prio v0.8.2 (/Users/chris/github.com/divviup/libprio-rs)
warning: unneeded `return` statement
   --> src/field.rs:263:9
    |
263 |         return false;
    |         ^^^^^^^^^^^^^ help: remove `return`: `false`
    |
    = note: `#[warn(clippy::needless_return)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: `prio` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 2.58s
```

MxmUrw commented:
> You've got one more thing to fix before we can merge.

Done.

tgeoghegan commented:
I'm going to review and merge this today. Thanks for your patience!

tgeoghegan commented:
Thanks for the contribution @MxmUrw!
